Resolves #355: Support grabbing the bytes of intervals

phadej opened the pull request.

phadej committed:
Resolves #355: Support grabbing the bytes of intervals

phadej commented:
I couldn't run macro benchmarks (#362), there are no differences in micro benchmarks

dcoutts reviewed (commented):
Thoughts so far.

Generally this looks very promising: providing the feature at low cost to non-users.

phadej commented on cborg/src/Codec/CBOR/Read.hs:
```diff
@@ -1419,25 +1506,41 @@ getTokenVarLen len bs offset =
-getTokenVarLenSlow :: [ByteString] -> Int -> ByteOffset
-                   -> IncrementalDecoder s (ByteString, ByteString)
-getTokenVarLenSlow bss n offset = do
+                         (offset + intToInt64 (BS.length bs'))
```
> I've not followed why we're changing the offset calculations here

Because `getTokenVarLen/Slow` didn't use `offset` for anything, only to report an error. Previously (as you can see from the diff) it wasn't updated at all even after new chunks were read.

A side-effect of this change is that "end-of-input" error will no report the actualy end-of-input offset, previously it reported an offset at which `getTokenVarLen` was called.

phadej commented on cborg/src/Codec/CBOR/Read.hs:
```diff
@@ -1250,15 +1318,15 @@ go_slow da bs !offset = do
       case mbs of
         Nothing   -> decodeFail bs' offset' "end of input"
-        Just bs'' -> go_slow da' bs'' offset'
+        Just bs'' -> go_slow da' bs'' offset' (slowMarkChunk bs'' offset' marks)
```
Well, I doubt there's any noticeable cost cause

```haskell
slowMarkChunk :: ByteString -> ByteOffset -> Marks -> Marks
slowMarkChunk _  _    NoMarks            = NoMarks
```

My gut feeling is there's more accumulated cost of threading an additional `NoMarks` argument through the interpreter.

dcoutts replied:
Indeed. And as you noted, the micro-benchmarks show nothing measureable. And I fixed the macro-benchmarks and they also show nothing measurable.

phadej commented on cborg/tests/Tests/ByteOffset.hs:
```diff
@@ -63,6 +66,24 @@ prop_ATerm_isomorphic3 t =
+    res = either throw snd (deserialiseFromBytes emptyDecoder (LBS.pack [0]))
+
+empty_deserialise_fail :: Bool
+empty_deserialise_fail = isLeft (deserialiseFromBytes emptyDecoder LBS.empty)
```
I added these tests for both `peekByteOffset` and `getInputSpan`. 

I find it a bit surprising as `peekByteOffset` as well as `unmarkInput` and `getInputSpan` do succeed at the end of input, but they seem to fail at the very beginning of an empty input.

This is quite obscure corner case, but nevertheless I think it's good to be aware off (i.e. have test for it).

dcoutts committed:
Change names of marks/input span to use "byte span" based names

Change:
markInput    --> openByteSpan
unmarkInput  --> closeByteSpan
getInputSpan --> peekByteSpan

dcoutts commented on cborg/src/Codec/CBOR/Read.hs:
```diff
@@ -1198,48 +1207,107 @@ go_fast_end !bs (ConsumeMapLenOrIndef k) =
 
     SlowConsumeTokenString bs' k len -> do
-      (bstr, bs'') <- getTokenVarLen len bs' offset'
+      (bstr, bs'', marks') <- getTokenVarLen len bs' (offset' + intToInt64 (BS.length bs')) marks
```
Got it. Thanks. And thanks for pushing the slight simplification.

dcoutts reviewed (approved):
This is great. It adds the feature nicely, and doesn't add any cost for users not using it. And indeed using the feature itself is pretty cheap (as measured by a macro-benchmark).

I've taken the liberty of doing some renaming. And I'll add more API docs. I'll merge once I've finished adding docs.

dcoutts committed:
Add docs for new byte span API

phadej commented on cborg/src/Codec/CBOR/Decoding.hs:
```diff
@@ -1047,7 +1125,7 @@ peekByteSpan = Decoder (\k -> return (PeekByteSpan k))
 -- > openByteSpan
 -- > x <- decode
--- > !after  <- peekByteSpan
+-- > bytes <- peekByteSpan
```
you removed bang here, but not in the implementation.

dcoutts replied:
Thanks. I think the bang there is unnecessary and it's better to force the call of `peekMarkedByteSpan` as it is passed to the continuation. I'll change that.

dcoutts committed:
Adjust where we evaluate selecting out the marked byte span.

Better to do it in the reader.

coot commented:
Are there any plans to release new version of `cborg` with this change included?
